plugins/ports-plugin/src/ports-plugin.ts

```diff
@@ -10,191 +10,326 @@
+    private showPluginEndpoints: boolean;
+
+    constructor(private context: theia.PluginContext) {
+        this.devfileEndpoints = [];
+        this.currentEndpoints = [];
+        this.redirectPorts = [];
+        this.showPluginEndpoints = false;
+        this.devfileHandler = new CheServerDevfileHandlerImpl();
+        this.redirectListeners = new Map();
+        this.excludedPorts = [];
+        this.endpointsTreeDataProvider = new EndpointsTreeDataProvider();
+        this.portChangesDetector = new PortChangesDetector();
+        this.outputChannel = theia.window.createOutputChannel('Ports Plug-in');
 
-    // grab a free redirect
-    if (redirectPorts.length === 0) {
-        await theia.window.showErrorMessage(errorMessage, { modal: true });
-        return;
     }
-
-    const interactions: MessageItem[] = [{ title: 'yes' }];
-    const result = await theia.window.showInformationMessage(redirectMessage, ...interactions);
-    if (result && result.title === 'yes') {
-        // takes first available port
-        const workspacePort = redirectPorts.pop()!;
-
-        // start a new listener
-        const portRedirectListener = new PortRedirectListener(parseInt(workspacePort.portNumber, 10), 'localhost', port.portNumber);
-        portRedirectListener.start();
-
-        // store port taken
-        const busyPort = { portRedirectListener, workspacePort };
-        redirectListeners.set(port.portNumber, busyPort);
-
-        // show redirect
-        const redirectInteractions: MessageItem[] = [{ title: 'Open Link' }];
-        const msg = `Redirect is now enabled on port ${port.portNumber}. External URL is ${workspacePort.url}`;
-        const resultShow = await theia.window.showInformationMessage(msg, ...redirectInteractions);
-        if (resultShow && resultShow.title === 'Open Link') {
-            theia.commands.executeCommand('mini-browser.openUrl', workspacePort.url);
+    /**
+     * Prompt user to create a port redirect for the specific port
+     * @param port the port that needs to be redirected
+     * @param redirectMessage the message if there are 'free ports' in workspace
+     * @param errorMessage  if no free port are available
+     */
+    async askRedirect(port: ListeningPort, redirectMessage: string, errorMessage: string): Promise<void> {
+
+        // grab a free redirect
+        if (this.redirectPorts.length === 0) {
+            await theia.window.showErrorMessage(errorMessage, { modal: true });
+            return;
         }
-    }
-}
-
-// Callback when a new port is being opened in workspace
-async function onOpenPort(port: Port): Promise<void> {
 
-    // skip excluded
-    if (excludedPorts.includes(port.portNumber)) {
-        // this port is excluded so just print a notice but does not propose a redirect
-        outputChannel.appendLine(`Ignoring excluded port ${port.portNumber}`);
-        return;
+        const interactions: MessageItem[] = [{ title: 'yes' }, { title: 'no' }];
```

Comment:
Yes but it was breaking happy path tests so I decided to make it a specific change done later for happy path enhancements to have this one being able to pass tests.

Reply:
Fair enough (even though I the software being correct should take precedence over satisfying the tests, IMO). In that case, there should be a "todo" comment referencing an issue against the happy path test. Otherwise, this won't get fixed later.